### PR TITLE
Added terraform provider for powerstore with volume operations

### PR DESCRIPTION
**Pull request by Sakshi-dell**
# Description
Moved terraform plugin for powerstore to terraform plugin framework from terraform SDKv2.
Added create, read and delete operations for Volume.

# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |
| |

# Checklist:

- [x] I have performed a self-review of my own code to ensure there are no formatting, vetting, linting, or security issues
- [ ] I have verified that new and existing unit tests pass locally with my changes
- [ ] I have not allowed coverage numbers to degenerate
- [ ] I have maintained at least 90% code coverage
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] Backward compatibility is not broken

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Please also list any relevant details for your test configuration

- [x] Created and destroyed volume with attributes

### CODE REVIEW

File: client/client.go [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+// Client type is to hold powerstore client
+type Client struct {
+	PStoreClient *pstore.ClientIMPL
+}
+
+var (
+	newClientWithArgs = pstore.NewClientWithArgs
+)
+
+// NewClient returns the gopowermax client
```
Comment (mkmayank-dell) [inline client/client.go]: its a typo, it supposed to be gopowerstore client in a comment

File: main.go [hunk 2]
```diff
@@ -1,17 +1,30 @@
...
-
+	"context"
+	"flag"
+	"github.com/hashicorp/terraform-plugin-framework/providerserver"
+	"log"
 	"terraform-provider-powerstore/powerstore"
 )
 
+var (
+	version string = "dev"
```
Comment (mkmayank-dell) [inline main.go:12]: shouldn't we make this as command line argument or may be a environment variable

**Review by mkmayank-dell — APPROVED** [review]

File: Makefile [hunk 3]
```diff
@@ -1,57 +1,27 @@
...
+	go generate
+
+download:
+	go mod download
+
+build: download
+	mkdir -p out
+	go build -v -o ./out
+
+all: download code-check test compile
```
Comment (shenda1) [inline Makefile]: test and compile available?

File: client/client.go (hunk 1 above)
Comment (shenda1) [inline client/client.go]: gopowerstore

File: examples/main.tf [hunk 4]
```diff
@@ -1,61 +1,33 @@
...
+  insecure = true
 }
 
-  resource "powerstore_storage_container" "test-sc" {
-  # (resource arguments)
-      name = "Test-SC_new"
-      quota = 109951162777
-}
\ No newline at end of file
+resource "powerstore_volume" "test" {
```
Comment (shenda1) [inline examples/main.tf:29]: Please add examples with all the supported parameters

File: powerstore/provider.go [hunk 5]
```diff
@@ -2,68 +2,236 @@ package powerstore
...
-		Schema: map[string]*schema.Schema{
-			"host": &schema.Schema{
-				Type:        schema.TypeString,
-				Optional:    true,
-				DefaultFunc: schema.EnvDefaultFunc("POWERSTORE_HOST", nil),
+type Pstoreprovider struct {
+	// client can contain the upstream provider SDK or HTTP client used to
+	// communicate with the upstream service. Resource and DataSource
+	// implementations can then make calls using this client.
+	//
```
Comment (shenda1) [inline powerstore/provider.go]: Please remove line if not required

File: powerstore/provider.go [hunk 6]
```diff
@@ -2,68 +2,236 @@ package powerstore
...
+	if config.Username.Unknown {
+		// Cannot connect to client with an unknown value
+		resp.Diagnostics.AddWarning(
+			"Unable to create client",
+			"Cannot use unknown value as username",
+		)
+		return
+	}
+
+	if config.Username.Value == "" {
```
Comment (shenda1) [inline powerstore/provider.go:63]: Please check if we need env variable other than acceptance tests. Applicable for other instances as well.

File: powerstore/provider.go [hunk 7]
```diff
@@ -2,68 +2,236 @@ package powerstore
...
+					DefaultAttribute(types.Bool{Value: false}),
+				},
 			},
-			"password": &schema.Schema{
-				Type:        schema.TypeString,
-				Optional:    true,
-				Sensitive:   true,
-				DefaultFunc: schema.EnvDefaultFunc("POWERSTORE_PASSWORD", nil),
+			"password": {
+				MarkdownDescription: "The password of the Unisphere host.",
```
Comment (shenda1) [inline powerstore/provider.go]: Unisphere?

File: powerstore/provider.go [hunk 8]
```diff
@@ -2,68 +2,236 @@ package powerstore
...
+				Description:         "The password of the Unisphere host.",
+				Type:                types.StringType,
+				Required:            true,
+				Sensitive:           true,
+			},
+			"username": {
+				MarkdownDescription: "The username of the Unisphere host.",
```
Comment (shenda1) [inline powerstore/provider.go]: Unisphere?

File: powerstore/volume_helper.go [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+	volState.Type.Value = string(volResponse.Type)
+	volState.WWN.Value = volResponse.Wwn
+	volState.Description.Value = volResponse.Description
+	volState.State.Value = string(volResponse.State)
+	volState.WWN.Value = volResponse.Wwn
+	volState.ApplianceID.Value = volResponse.ApplianceID
+	volState.CreationTimeStamp.Value = volResponse.CreationTimeStamp
+	volState.ProtectionPolicyID.Value = volResponse.ProtectionPolicyID
+	volState.PerformancePolicyID.Value = volResponse.PerformancePolicyID
+	if len(hostMapping) > 0 {
```
Comment (shenda1) [inline powerstore/volume_helper.go:25]: Please add the comments

File: client/client.go (hunk 1 above)
Comment (Sakshi-dell) [reply client/client.go]: Updated

File: Makefile (hunk 3 above)
Comment (Sakshi-dell) [reply Makefile]: Updated

File: examples/main.tf (hunk 4 above)
Comment (Sakshi-dell) [reply examples/main.tf:29]: Added all the fields

File: powerstore/provider.go (hunk 6 above)
Comment (Sakshi-dell) [reply powerstore/provider.go:63]: Removed env variables

File: examples/main.tf [hunk 10]
```diff
@@ -1,61 +1,43 @@
...
+  name = "test_vol"
+  size = 7516192768
+  description = "Creating volume"
+  host_id=""
+  host_group_id=""
+  appliance_id="A1"
+  volume_group_id=""
```
Comment (shenda1) [inline examples/main.tf:36]: Do we need volume_group_id?

Comment (Sakshi-dell) [reply examples/main.tf:36]: Yes. Volume_group_ID is a part of the volume create instance, It can also be mapped and unmapped during volume update.

**Review by shenda1 — APPROVED** [review]

**Pull request merged**